The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6646438264?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/69627f5e786dba7b69d9e37b0ea58f9f739070f9/checks/6646438264/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
........................................................................................ 6952/13290
....................................i....i.........................................i.... 7040/13290
.............i.............i.......................................................i.... 7128/13290
........................................................................................ 7216/13290
F.F.........i.......................................................F................... 7304/13290
.....................................................................................ii. 7480/13290
....................................ii.................................................. 7568/13290
.........i.............................................................................. 7656/13290
........................................................................................ 7744/13290
---
........................................................................................ 11000/13290
........................................................................................ 11088/13290
........................................................................................ 11176/13290
........................................................................................ 11264/13290
.....................F...FF............................................................. 11352/13290
........................................................................................ 11528/13290
........................................................................................ 11616/13290
........................................................................................ 11704/13290
..............................................................i........i........i.....i. 11792/13290
---

8    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
9    = note: for more information, see issue #64266 <https://github.com/rust-lang/rust/issues/64266>
10 
- error: use of unstable library feature 'test': `bench` is a part of custom test frameworks which are unstable
-   --> $DIR/bench.rs:7:5
-    |
- LL | use bench as _;
-    |
-    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
-    = note: for more information, see issue #64266 <https://github.com/rust-lang/rust/issues/64266>
- 
---
To only update this specific test, also pass `--test-args feature-gates/bench.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/feature-gates/bench.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/bench" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2018" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/bench/auxiliary"
stdout: none
--- stderr -------------------------------
error: use of unstable library feature 'test': `bench` is a part of custom test frameworks which are unstable
   |
   |
LL | #[bench] //~ ERROR use of unstable library feature 'test'
   |
   = note: `#[deny(soft_unstable)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #64266 <https://github.com/rust-lang/rust/issues/64266>
---
---- [ui] src/test/ui/feature-gates/issue-49983-see-issue-0.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/feature-gates/issue-49983-see-issue-0.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/issue-49983-see-issue-0" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/issue-49983-see-issue-0/auxiliary"
stdout: none
stderr: none

---- [ui] src/test/ui/issues/issue-52489.rs stdout ----

error: ui test compiled successfully!
error: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/issues/issue-52489.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-52489" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2018" "--extern" "issue_52489" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/issues/issue-52489/auxiliary"
stdout: none
stderr: none

---- [ui] src/test/ui/lint/lint-output-format-2.rs stdout ----
diff of stderr:


1 warning: use of deprecated function `lint_output_format::foo`: text
-   --> $DIR/lint-output-format-2.rs:7:26
-    |
- LL | use lint_output_format::{foo, bar};
-    |
-    = note: `#[warn(deprecated)]` on by default
- 
- 
- warning: use of deprecated function `lint_output_format::foo`: text
11    |
12 LL |     let _x = foo();

13    |              ^^^
---
17 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-output-format-2/lint-output-format-2.stderr
To only update this specific test, also pass `--test-args lint/lint-output-format-2.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/lint-output-format-2.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-output-format-2" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-output-format-2/auxiliary"
stdout: none
--- stderr -------------------------------
warning: use of deprecated function `lint_output_format::foo`: text
   |
LL |     let _x = foo();
   |              ^^^
   |
---
8 
9 error[E0658]: use of unstable library feature 'unstable_test_feature'
-   --> $DIR/lint-output-format.rs:7:31
-    |
- LL | use lint_output_format::{foo, bar};
-    |
-    = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable
- 
- error[E0658]: use of unstable library feature 'unstable_test_feature'
- error[E0658]: use of unstable library feature 'unstable_test_feature'
18   --> $DIR/lint-output-format.rs:11:14
19    |
20 LL |     let _y = bar();
22    |
23    = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable
24 
- error: aborting due to 3 previous errors
---
To only update this specific test, also pass `--test-args lint/lint-output-format.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/lint-output-format.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-output-format" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-F" "unused_features" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-output-format/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0658]: use of unstable library feature 'unstable_test_feature'
   |
   |
LL | extern crate lint_output_format; //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-output-format.rs:11:14
   |
LL |     let _y = bar(); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error: aborting due to 2 previous errors
---
To only update this specific test, also pass `--test-args lint/lint-stability.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/lint-stability.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-stability" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-stability/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0658]: use of unstable library feature 'unstable_test_feature'
   |
   |
LL |     extern crate stability_cfg2; //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
---

error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:47:9
   |
LL |         Trait::trait_deprecated_unstable(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:49:9
   |
LL |         <Foo as Trait>::trait_deprecated_unstable(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
---

error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:54:9
   |
LL |         Trait::trait_deprecated_unstable_text(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:56:9
   |
LL |         <Foo as Trait>::trait_deprecated_unstable_text(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:59:9
   |
LL |         unstable(); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:60:9
   |
LL |         Trait::trait_unstable(&foo); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:61:9
   |
LL |         <Foo as Trait>::trait_unstable(&foo); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable


error[E0658]: use of unstable library feature 'unstable_test_feature': text
   |
LL |         unstable_text();
   |         ^^^^^^^^^^^^^
   |
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature': text
   |
   |
LL |         Trait::trait_unstable_text(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable


error[E0658]: use of unstable library feature 'unstable_test_feature': text
   |
   |
LL |         <Foo as Trait>::trait_unstable_text(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:99:17
   |
LL |         let _ = DeprecatedUnstableStruct {
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:103:17
   |
LL |         let _ = UnstableStruct { i: 0 }; //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:107:17
   |
LL |         let _ = DeprecatedUnstableUnitStruct;
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:109:17
   |
LL |         let _ = UnstableUnitStruct; //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:113:17
   |
LL |         let _ = Enum::DeprecatedUnstableVariant;
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:115:17
   |
LL |         let _ = Enum::UnstableVariant; //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:119:17
   |
LL |         let _ = DeprecatedUnstableTupleStruct (1);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:121:17
   |
LL |         let _ = UnstableTupleStruct (1); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:130:25
   |
LL |         macro_test_arg!(deprecated_unstable_text());
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:144:9
   |
LL |         Trait::trait_deprecated_unstable(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:146:9
   |
LL |         <Foo as Trait>::trait_deprecated_unstable(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:148:9
   |
LL |         Trait::trait_deprecated_unstable_text(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:150:9
   |
LL |         <Foo as Trait>::trait_deprecated_unstable_text(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:152:9
   |
LL |         Trait::trait_unstable(&foo); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:153:9
   |
LL |         <Foo as Trait>::trait_unstable(&foo); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable


error[E0658]: use of unstable library feature 'unstable_test_feature': text
   |
   |
LL |         Trait::trait_unstable_text(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable


error[E0658]: use of unstable library feature 'unstable_test_feature': text
   |
   |
LL |         <Foo as Trait>::trait_unstable_text(&foo);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:172:10
   |
LL |     impl UnstableTrait for S { } //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:174:24
   |
LL |     trait LocalTrait : UnstableTrait { } //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:179:9
   |
LL |         fn trait_unstable(&self) {} //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:184:5
   |
LL |     extern crate inherited_stability; //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:188:9
   |
LL |         unstable(); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:191:9
   |
LL |         stable_mod::unstable(); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:195:9
   |
LL |         unstable_mod::unstable(); //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:197:17
   |
LL |         let _ = Unstable::UnstableVariant; //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:88:48
   |
LL |         struct S1<T: TraitWithAssociatedTypes>(T::TypeUnstable);
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'unstable_test_feature'
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/lint/lint-stability.rs:92:13
   |
LL |             TypeUnstable = u8, //~ ERROR use of unstable library feature
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error: aborting due to 40 previous errors
---
---- [ui] src/test/ui/stability-attribute/issue-28075.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/stability-attribute/issue-28075.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/stability-attribute/issue-28075" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/stability-attribute/issue-28075/auxiliary"
stdout: none
stderr: none

---- [ui] src/test/ui/stability-attribute/issue-28388-3.rs stdout ----

error: ui test compiled successfully!
error: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/stability-attribute/issue-28388-3.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/stability-attribute/issue-28388-3" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/stability-attribute/issue-28388-3/auxiliary"
stdout: none
stderr: none


---- [ui] src/test/ui/stability-attribute/allow-unstable-reexport.rs stdout ----

1 error[E0658]: use of unstable library feature 'unstable_test_feature'
-   --> $DIR/allow-unstable-reexport.rs:17:5
+   --> $DIR/allow-unstable-reexport.rs:19:5
+   --> $DIR/allow-unstable-reexport.rs:19:5
3    |
4 LL |     unstable();
5    |     ^^^^^^^^

7    = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable
8 
9 error[E0658]: use of unstable library feature 'unstable_test_feature': text
-   --> $DIR/allow-unstable-reexport.rs:18:5
11    |
12 LL |     unstable_text();
13    |     ^^^^^^^^^^^^^

---
To only update this specific test, also pass `--test-args stability-attribute/allow-unstable-reexport.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/stability-attribute/allow-unstable-reexport.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/stability-attribute/allow-unstable-reexport" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/stability-attribute/allow-unstable-reexport/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0658]: use of unstable library feature 'unstable_test_feature'
  --> /checkout/src/test/ui/stability-attribute/allow-unstable-reexport.rs:19:5
   |
LL |     unstable(); //~ ERROR use of unstable library feature 'unstable_test_feature'
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable


error[E0658]: use of unstable library feature 'unstable_test_feature': text
  --> /checkout/src/test/ui/stability-attribute/allow-unstable-reexport.rs:20:5
   |
LL |     unstable_text(); //~ ERROR use of unstable library feature 'unstable_test_feature'
   |
   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable

error: aborting due to 2 previous errors
---

---- [ui] src/test/ui/traits/issue-78372.rs stdout ----
diff of stderr:

35    |       help: you might be missing a type parameter: `, MISC`
37 error[E0658]: use of unstable library feature 'dispatch_from_dyn'
-   --> $DIR/issue-78372.rs:1:5
-    |
- LL | use std::ops::DispatchFromDyn;
- LL | use std::ops::DispatchFromDyn;
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^
-    |
-    = help: add `#![feature(dispatch_from_dyn)]` to the crate attributes to enable
- error[E0658]: use of unstable library feature 'dispatch_from_dyn'
46   --> $DIR/issue-78372.rs:3:9
47    |
47    |
48 LL | impl<T> DispatchFromDyn<Smaht<U, MISC>> for T {}

65 LL | impl<T> DispatchFromDyn<Smaht<U, MISC>> for T {}
67 
- error: aborting due to 7 previous errors
+ error: aborting due to 6 previous errors
69 
---
To only update this specific test, also pass `--test-args traits/issue-78372.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/traits/issue-78372.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-78372" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/traits/issue-78372/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0412]: cannot find type `PhantomData` in this scope
   |
   |
LL | struct Smaht<T, MISC>(PhantomData); //~ ERROR cannot find type `PhantomData` in this scope
   |
help: consider importing this struct
   |
LL | use std::marker::PhantomData;
LL | use std::marker::PhantomData;
   |

error[E0412]: cannot find type `U` in this scope
  --> /checkout/src/test/ui/traits/issue-78372.rs:3:31
   |
LL | impl<T> DispatchFromDyn<Smaht<U, MISC>> for T {} //~ ERROR cannot find type `U` in this scope
   |      -                        ^
   |      similarly named type parameter `T` defined here
   |
help: a type parameter with a similar name exists
   |
   |
LL | impl<T> DispatchFromDyn<Smaht<T, MISC>> for T {} //~ ERROR cannot find type `U` in this scope
help: you might be missing a type parameter
   |
   |
LL | impl<T, U> DispatchFromDyn<Smaht<U, MISC>> for T {} //~ ERROR cannot find type `U` in this scope

error[E0412]: cannot find type `MISC` in this scope
  --> /checkout/src/test/ui/traits/issue-78372.rs:3:34
   |
   |
LL | impl<T> DispatchFromDyn<Smaht<U, MISC>> for T {} //~ ERROR cannot find type `U` in this scope
   |       -                          ^^^^ not found in this scope
   |       |
   |       help: you might be missing a type parameter: `, MISC`
error[E0658]: use of unstable library feature 'dispatch_from_dyn'
  --> /checkout/src/test/ui/traits/issue-78372.rs:3:9
   |
   |
LL | impl<T> DispatchFromDyn<Smaht<U, MISC>> for T {} //~ ERROR cannot find type `U` in this scope
   |
   |
   = help: add `#![feature(dispatch_from_dyn)]` to the crate attributes to enable

error[E0210]: type parameter `T` must be covered by another type when it appears before the first local type (`Smaht<[type error], [type error]>`)
   |
   |
LL | impl<T> DispatchFromDyn<Smaht<U, MISC>> for T {} //~ ERROR cannot find type `U` in this scope
   |      ^ type parameter `T` must be covered by another type when it appears before the first local type (`Smaht<[type error], [type error]>`)
   |
   = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local, and no uncovered type parameters appear before that first local type
   = note: in this case, 'before' refers to the following order: `impl<..> ForeignTrait<T1, ..., Tn> for T0`, where `T0` is the first and `Tn` is the last

error[E0378]: the trait `DispatchFromDyn` may only be implemented for a coercion between structures
   |
   |
LL | impl<T> DispatchFromDyn<Smaht<U, MISC>> for T {} //~ ERROR cannot find type `U` in this scope

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0210, E0378, E0412, E0658.
```

</details>

--- a/compiler/rustc_middle/src/middle/stability.rs
+++ b/compiler/rustc_middle/src/middle/stability.rs
@@ -9,10 +9,9 @@ use rustc_attr::{self as attr, ConstStability, Deprecation, Stability};
 use rustc_data_structures::fx::FxHashMap;
 use rustc_errors::{Applicability, Diagnostic};
 use rustc_feature::GateIssue;
-use rustc_hir as hir;
 use rustc_hir::def::DefKind;
 use rustc_hir::def_id::{DefId, LocalDefId};
-use rustc_hir::{self, HirId};
+use rustc_hir::{self as hir, HirId};
 use rustc_middle::ty::print::with_no_trimmed_paths;
 use rustc_session::lint::builtin::{DEPRECATED, DEPRECATED_IN_FUTURE, SOFT_UNSTABLE};
 use rustc_session::lint::{BuiltinLintDiagnostics, Level, Lint, LintBuffer};
@@ -306,6 +305,14 @@ fn suggestion_for_allocator_api(
     None
 }
 
+/// An override option for eval_stability.
+pub enum AllowUnstable {
+    /// Don't emit an unstable error for the item
+    Yes,
+    /// Handle the item normally
+    No,
+}
+
 impl<'tcx> TyCtxt<'tcx> {
     /// Evaluates the stability of an item.
     ///
@@ -322,6 +329,28 @@ impl<'tcx> TyCtxt<'tcx> {
         id: Option<HirId>,
         span: Span,
         method_span: Option<Span>,
+    ) -> EvalResult {
+        self.eval_stability_allow_unstable(def_id, id, span, method_span, AllowUnstable::No)
+    }
+
+    /// Evaluates the stability of an item.
+    ///
+    /// Returns `EvalResult::Allow` if the item is stable, or unstable but the corresponding
+    /// `#![feature]` has been provided. Returns `EvalResult::Deny` which describes the offending
+    /// unstable feature otherwise.
+    ///
+    /// If `id` is `Some(_)`, this function will also check if the item at `def_id` has been
+    /// deprecated. If the item is indeed deprecated, we will emit a deprecation lint attached to
+    /// `id`.
+    ///
+    /// Pass `AllowUnstable::Yes` to `allow_unstable` to force an unstable item to be allowed. Deprecation warnings will be emitted normally.
+    pub fn eval_stability_allow_unstable(
+        self,
+        def_id: DefId,
+        id: Option<HirId>,
+        span: Span,
+        method_span: Option<Span>,
+        allow_unstable: AllowUnstable,
     ) -> EvalResult {
         // Deprecated attributes apply in-crate and cross-crate.
         if let Some(id) = id {
@@ -419,6 +448,10 @@ impl<'tcx> TyCtxt<'tcx> {
                     }
                 }
 
+                if matches!(allow_unstable, AllowUnstable::Yes) {
+                    return EvalResult::Allow;
+                }
+
                 let suggestion = suggestion_for_allocator_api(self, def_id, span, feature);
                 EvalResult::Deny { feature, reason, issue, suggestion, is_soft }
             }
@@ -445,11 +478,38 @@ impl<'tcx> TyCtxt<'tcx> {
         span: Span,
         method_span: Option<Span>,
     ) {
-        self.check_optional_stability(def_id, id, span, method_span, |span, def_id| {
-            // The API could be uncallable for other reasons, for example when a private module
-            // was referenced.
-            self.sess.delay_span_bug(span, &format!("encountered unmarked API: {:?}", def_id));
-        })
+        self.check_stability_allow_unstable(def_id, id, span, method_span, AllowUnstable::No)
+    }
+
+    /// Checks if an item is stable or error out.
+    ///
+    /// If the item defined by `def_id` is unstable and the corresponding `#![feature]` does not
+    /// exist, emits an error.
+    ///
+    /// This function will also check if the item is deprecated.
+    /// If so, and `id` is not `None`, a deprecated lint attached to `id` will be emitted.
+    ///
+    /// Pass `AllowUnstable::Yes` to `allow_unstable` to force an unstable item to be allowed. Deprecation warnings will be emitted normally.
+    pub fn check_stability_allow_unstable(
+        self,
+        def_id: DefId,
+        id: Option<HirId>,
+        span: Span,
+        method_span: Option<Span>,
+        allow_unstable: AllowUnstable,
+    ) {
+        self.check_optional_stability(
+            def_id,
+            id,
+            span,
+            method_span,
+            allow_unstable,
+            |span, def_id| {
+                // The API could be uncallable for other reasons, for example when a private module
+                // was referenced.
+                self.sess.delay_span_bug(span, &format!("encountered unmarked API: {:?}", def_id));
+            },
+        )
     }
 
     /// Like `check_stability`, except that we permit items to have custom behaviour for
@@ -462,6 +522,7 @@ impl<'tcx> TyCtxt<'tcx> {
         id: Option<HirId>,
         span: Span,
         method_span: Option<Span>,
+        allow_unstable: AllowUnstable,
         unmarked: impl FnOnce(Span, DefId),
     ) {
         let soft_handler = |lint, span, msg: &_| {
@@ -469,7 +530,7 @@ impl<'tcx> TyCtxt<'tcx> {
                 lint.build(msg).emit();
             })
         };
-        match self.eval_stability(def_id, id, span, method_span) {
+        match self.eval_stability_allow_unstable(def_id, id, span, method_span, allow_unstable) {
             EvalResult::Allow => {}
             EvalResult::Deny { feature, reason, issue, suggestion, is_soft } => report_unstable(
                 self.sess,

--- a/compiler/rustc_passes/src/stability.rs
+++ b/compiler/rustc_passes/src/stability.rs
@@ -9,10 +9,10 @@ use rustc_hir::def::{DefKind, Res};
 use rustc_hir::def_id::{LocalDefId, CRATE_DEF_ID};
 use rustc_hir::hir_id::CRATE_HIR_ID;
 use rustc_hir::intravisit::{self, Visitor};
-use rustc_hir::{FieldDef, Generics, HirId, Item, TraitRef, Ty, TyKind, Variant};
+use rustc_hir::{FieldDef, Generics, HirId, Item, ItemKind, TraitRef, Ty, TyKind, Variant};
 use rustc_middle::hir::nested_filter;
 use rustc_middle::middle::privacy::AccessLevels;
-use rustc_middle::middle::stability::{DeprecationEntry, Index};
+use rustc_middle::middle::stability::{AllowUnstable, DeprecationEntry, Index};
 use rustc_middle::ty::{self, query::Providers, TyCtxt};
 use rustc_session::lint;
 use rustc_session::lint::builtin::{INEFFECTIVE_UNSTABLE_TRAIT_IMPL, USELESS_DEPRECATED};
@@ -807,10 +807,44 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
     fn visit_path(&mut self, path: &'tcx hir::Path<'tcx>, id: hir::HirId) {
         if let Some(def_id) = path.res.opt_def_id() {
             let method_span = path.segments.last().map(|s| s.ident.span);
-            self.tcx.check_stability(def_id, Some(id), path.span, method_span)
+            self.tcx.check_stability_allow_unstable(
+                def_id,
+                Some(id),
+                path.span,
+                method_span,
+                if is_unstable_reexport(self.tcx, id) {
+                    AllowUnstable::Yes
+                } else {
+                    AllowUnstable::No
+                },
+            )
         }
         intravisit::walk_path(self, path)
     }
+}
+
+/// Check whether a path is a `use` item that has been marked as unstable.
+///
+/// See issue #94972 for details on why this is a special case
+fn is_unstable_reexport<'tcx>(tcx: TyCtxt<'tcx>, id: hir::HirId) -> bool {
+    // Get the LocalDefId so we can lookup the item to check the kind.
+    let Some(def_id) = tcx.hir().opt_local_def_id(id) else { return false; };
+
+    let Some(stab) = tcx.stability().local_stability(def_id) else {
+        return false;
+    };
+
+    if stab.level.is_stable() {
+        // The re-export is not marked as unstable, don't override
+        return false;
+    }
+
+    // If this is a path that isn't a use, we don't need to do anything special
+    if !matches!(tcx.hir().item(hir::ItemId { def_id }).kind, ItemKind::Use(..)) {
+        return false;
+    }
+
+    true
 }
 
 struct CheckTraitImplStable<'tcx> {

--- a/compiler/rustc_typeck/src/astconv/mod.rs
+++ b/compiler/rustc_typeck/src/astconv/mod.rs
@@ -24,6 +24,7 @@ use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_hir::intravisit::{walk_generics, Visitor as _};
 use rustc_hir::lang_items::LangItem;
 use rustc_hir::{GenericArg, GenericArgs};
+use rustc_middle::middle::stability::AllowUnstable;
 use rustc_middle::ty::subst::{self, GenericArgKind, InternalSubsts, Subst, SubstsRef};
 use rustc_middle::ty::GenericParamDefKind;
 use rustc_middle::ty::{self, Const, DefIdTree, EarlyBinder, Ty, TyCtxt, TypeFoldable};
@@ -426,6 +427,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                             Some(arg.id()),
                             arg.span(),
                             None,
+                            AllowUnstable::No,
                             |_, _| {
                                 // Default generic parameters may not be marked
                                 // with stability attributes, i.e. when the

--- a/src/test/ui/stability-attribute/allow-unstable-reexport.rs
+++ b/src/test/ui/stability-attribute/allow-unstable-reexport.rs
@@ -1,0 +1,30 @@
+// Allow an unstable re-export without requiring a feature gate.
+// #94972
+
+// aux-build:lint-stability.rs
+// aux-build:lint-stability-reexport.rs
+#![feature(staged_api)]
+#![stable(feature = "lint_stability", since = "1.0.0")]
+
+extern crate lint_stability;
+extern crate lint_stability_reexport;
+
+#[unstable(feature = "unstable_test_feature", issue = "none")]
+pub use lint_stability::unstable;
+
+// We want to confirm that using a re-export through another crate behaves
+// the same way as using an item directly
+#[unstable(feature = "unstable_test_feature", issue = "none")]
+pub use lint_stability_reexport::unstable_text;
+
+// Ensure items which aren't marked as unstable can't re-export unstable items
+#[stable(feature = "lint_stability", since = "1.0.0")]
+pub use lint_stability::unstable as unstable2;
+//~^ ERROR use of unstable library feature 'unstable_test_feature'
+
+fn main() {
+    // Since we didn't enable the feature in this crate, we still can't
+    // use these items, even though they're in scope from the `use`s which are now allowed.
+    unstable(); //~ ERROR use of unstable library feature 'unstable_test_feature'
+    unstable_text(); //~ ERROR use of unstable library feature 'unstable_test_feature'
+}

--- a/src/test/ui/stability-attribute/allow-unstable-reexport.stderr
+++ b/src/test/ui/stability-attribute/allow-unstable-reexport.stderr
@@ -1,0 +1,27 @@
+error[E0658]: use of unstable library feature 'unstable_test_feature'
+  --> $DIR/allow-unstable-reexport.rs:22:9
+   |
+LL | pub use lint_stability::unstable as unstable2;
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable
+
+error[E0658]: use of unstable library feature 'unstable_test_feature'
+  --> $DIR/allow-unstable-reexport.rs:28:5
+   |
+LL |     unstable();
+   |     ^^^^^^^^
+   |
+   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable
+
+error[E0658]: use of unstable library feature 'unstable_test_feature': text
+  --> $DIR/allow-unstable-reexport.rs:29:5
+   |
+LL |     unstable_text();
+   |     ^^^^^^^^^^^^^
+   |
+   = help: add `#![feature(unstable_test_feature)]` to the crate attributes to enable
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0658`.

--- a/src/test/ui/stability-attribute/auxiliary/lint-stability-reexport.rs
+++ b/src/test/ui/stability-attribute/auxiliary/lint-stability-reexport.rs
@@ -1,0 +1,9 @@
+#![crate_type = "lib"]
+#![feature(staged_api)]
+#![stable(feature = "lint_stability", since = "1.0.0")]
+
+extern crate lint_stability;
+
+// Re-exporting without enabling the feature "unstable_test_feature" in this crate
+#[unstable(feature = "unstable_test_feature", issue = "none")]
+pub use lint_stability::unstable_text;

